src/libstd/sys/unix/mutex.rs

```diff
@@ -67,3 +68,45 @@ impl Mutex {
         debug_assert!(r == 0 || r == libc::EINVAL);
     }
 }
+
+pub struct ReentrantMutex { inner: Box<UnsafeCell<ffi::pthread_mutex_t>> }
```

Comment:
Could you leave a `FIXME` here to remove the `Box` at some point? Right now it unfortunately boxes twice (once at the common layer and once here as well). Fine for now, just want to make sure we don't forget to remove it at some point (these are supposed to be the lowest level abstractions).